### DISCUSSION TITLE
Matching users after import

### DISCUSSION THREAD

**Discussion by mipimipi** [umputun/remark42#915] [Q&A]
I imported lots of comments into remark42 from a Drupal web site using the Disqus format, incl. users. It seems that these users got id's "disqus_...". Now, when such a user log into remark42 using authentication via mail, a new userid is generated "email_...". Is there any possibility to match the user ids? I want to achieve that if a user logs in after the import, his name/id is match to the user name/id of the corresponding imported comments. Is that possible? Thanks!

**Comment by umputun** [comment]
I don't see what we will be matching on. All we got after import from the disqus is the user name. Email is not stored anywhere as we don't want to keep emails in the system. So, it leaves only one thing for the match - name. Obviously, this is not unique enough to produce reliable pairs.

Theoretically, we can keep the email from Disqus somewhere. I don't like this idea, but let say we do it. I still not sure how this flow supposed to work because we don't know anything about this user in remark42 until he posts a comment. Only at this moment, we have some clue, i.e. user_id. Pls note - we don't have user email and the only emails we can access are those submitted explicitly to get update notifications. As you can see not much to match on this side either. 

Another issue - what about all other login methods? Auth providers (on our side) designed with the lowest possible scope in mind and in many cases don't have access to emails, and even if they do - in those cases we not passing emails and not saving them anywhere. 

The more promising idea will be to allow the user to link ids. In this case, somehow user has to prove disqus identity and we could run some batch updates for all comments and normalize them to remark42 ids. I have no clue how to confirm this identity, just an idea.

**Reply by mipimipi** [reply]
Thanks. I understand the problem and expected that :-(. I assume that every time I log in via mail authentication, my user gets a new id, correct? What about the other Auth providers (Facebook, Twitter etc.)? Here, I hope that every time I log in, my user gets the same id, correct?

**Reply by umputun** [reply]
No, the email auth generates the [same ID](https://github.com/go-pkgz/auth/blob/master/provider/verify.go#L92) for the same email. Other providers (except anonymous) work the same. In other words, every time a user login with the same method he gets the same id

**Reply by mipimipi** [reply]
OK, understand. The userid of a user that authenticated via mail is stable since the part after "email_" is the hash of the mail address. 
Actually, I need the matching of the userids (what I mentioned above) just for the admins (just a handful of people). I will try to take a backup file, replace the disqus userids for the admin by their email userid and import the file. BTW: At the very beginning of the backup file I see the row `{"version":1,"users":[],"posts":[]}`. What is the purpose of the users and the posts array?

**Reply by umputun** [reply]
>  replace the disqus userids for the admin by their email userid and import the file

this should work, yes

> What is the purpose of the users and the posts array?

You don't care about it for disqus import. The goal of those elements to provide some [meta info](https://github.com/umputun/remark42/blob/682758a58ec50642858edb2c92e52d7410aa3718/backend/app/store/service/service.go#L54) about users (for example blocked users) and [posts](https://github.com/umputun/remark42/blob/682758a58ec50642858edb2c92e52d7410aa3718/backend/app/store/service/service.go#L65) (for example posts with disabling comments). It used by backup/restore, which is technically almost the same as the import.

**Reply by mipimipi** [reply]
The replacement of the userids in the backup file followed by a restore worked. The imported users are now matched with the user who logged in after the import :-) Thanks!
